sklearn/ensemble/_forest.py

```diff
@@ -803,52 +894,65 @@ def predict(self, X):
-        )
-
-        for estimator in self.estimators_:
-            unsampled_indices = _generate_unsampled_indices(
-                estimator.random_state, n_samples, n_samples_bootstrap)
-            p_estimator = estimator.predict(
-                X[unsampled_indices, :], check_input=False)
-
-            if self.n_outputs_ == 1:
-                p_estimator = p_estimator[:, np.newaxis]
-
-            predictions[unsampled_indices, :] += p_estimator
-            n_predictions[unsampled_indices, :] += 1
+        Parameters
+        ----------
+        tree : DecisionTreeRegressor object
+            A single decision tree regressor.
+        X : ndarray of shape (n_samples, n_features)
+            The OOB samples.
 
-        if (n_predictions == 0).any():
-            warn("Some inputs do not have OOB scores. "
-                 "This probably means too few trees were used "
-                 "to compute any reliable oob estimates.")
-            n_predictions[n_predictions == 0] = 1
+        Returns
+        -------
+        y_pred : ndarray of shape (n_samples, 1, n_outputs)
+            The OOB associated predictions.
+        """
+        y_pred = tree.predict(X, check_input=False)
+        if y_pred.ndim == 1:
+            # single output regression
+            y_pred = y_pred[:, np.newaxis, np.newaxis]
+        else:
+            # multioutput regression
+            y_pred = y_pred[:, np.newaxis, :]
+        return y_pred
 
-        predictions /= n_predictions
-        self.oob_prediction_ = predictions
+    @staticmethod
+    def _score_oob_predictions(y_true, oob_pred):
+        """Compute the average R2 score for the OOB samples.
 
-        if self.n_outputs_ == 1:
-            self.oob_prediction_ = \
-                self.oob_prediction_.reshape((n_samples, ))
+        Parameters
+        ----------
+        y_true : ndarray of shape (n_samples,) or (n_samples, n_outputs)
+            The true target.
+        oob_pred : ndarray of shape (n_samples, 1, n_outputs)
+            The OOB predictions.
 
-        self.oob_score_ = 0.0
+        Returns
+        -------
+        oob_score : float
+            The average accuracy score.
+        """
+        return r2_score(y_true, oob_pred[:, 0, :])
 
-        for k in range(self.n_outputs_):
-            self.oob_score_ += r2_score(y[:, k],
-                                        predictions[:, k])
+    def _set_oob_score_and_attributes(self, X, y):
+        """Compute and set the OOB score and attributes.
 
-        self.oob_score_ /= self.n_outputs_
+        Parameters
+        ----------
+        X : array-like of shape (n_samples, n_features)
+            The data matrix.
+        y : ndarray of shape (n_samples, n_outputs)
+            The target matrix.
+        """
+        oob_pred = super()._compute_oob_predictions(X, y)
+        self.oob_score_ = self._score_oob_predictions(y, oob_pred)
+        self.oob_prediction_ = oob_pred.squeeze(axis=1)
```

Comment:
Do we really need to introduce the `_score_oob_predictions` helper method? It feels like an unnecessary indirection. We could just do the following:

```suggestion
        self.oob_prediction_ = oob_pred.squeeze(axis=1)
        self.oob_score_ = r2_score(y_true, self.oob_prediction_)
```

and similarly for the classifiers with `accuracy_score`.

Reply:
Oh I see, did not got it at first.

Reply:
We might reintroduce it in the OOB permutation importance but we can see in the PR.